### ISSUE TITLE
Codec#register is called from the initialize method

### ISSUE THREAD

**Issue by jakelandis** [elastic/logstash#9108]
Codecs are modeled very similar to the inputs, filters and outputs. However, the the `register` method is for codecs is optional, and called from it's own initialize method. 

This is an odd design that such that it differs from the other plugins.  Further, this can lead to bad assumptions w/r/t the differences between class initialization and plugin registration.   The plugin classes are initialized to check it's setting to see if it is re-loadable, is an example of when you want to initialize but not register.  

I believe that the semantic difference between initialize and register is that you can initialize anytime (for example to check settings),  but only register when you want to use the business logic provided by the plugin.  Codec's blur that semantic difference. 

I would suggest that we add `register` to the plugin parent class with documentation to it's semantic purpose, and remove calling codec#register from codec#initialize in favor of letting the pipeline do that work.

**Comment by jakelandis** [comment]
Thanks @IrlJidel for raising awareness of this on  https://github.com/elastic/logstash/issues/8902#issuecomment-361566078

**Comment by andrewvc** [comment]
I agree that this is weird, and makes no sense. We should initialize these in the pipeline.

**Comment by guyboertje** [comment]
This was changed to support arbitrary cloning of codecs when there is a requirement to have a codec per data source stream.

IIRC there was/is no way of cloning a codec with pre-validated params.
